dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/runtime.hpp

```diff
@@ -1,113 +1,146 @@
+/** @file runtime.hpp
+*/
+
 #ifndef RUNTIME_HPP_INCLUDED
 #define RUNTIME_HPP_INCLUDED
 
 #include <thread>
 #include <atomic>
 #include <mutex>
 #include <condition_variable>
-#include <vector>
 
-#include "runtimetypes.hpp"
 #include "istatemachine.hpp"
 #include "threadpool.hpp"
 #include "ievent.hpp"
 #include "threadpoolmanager.hpp"
-#include "runtimetypes.hpp"
+#include "ESRoot/Types.hpp"
+#include "ESRoot/Containers/FixedArray.hpp"
+
+namespace Execution 
+{
 
 template<typename RuntimeType>
-class RuntimeI
+class IRuntime
 {
 public:
 
-  static RuntimeI<RuntimeType>* createRuntime()
-  {
-      return RuntimeType::createRuntime();
-  }
-
-  void setupObject(IStateMachine* sm_)
-  {
-      static_cast<RuntimeType*>(this)->setupObjectSpecificRuntime(sm_);
-  }
-  
-  void enqueueObject(IStateMachine* sm)
-  {
-      static_cast<RuntimeType*>(this)->enqueueObject(sm);
-  }
-  
-  void configure(ThreadConfiguration* configuration)
-  {
-	  if(!(static_cast<RuntimeType*>(this)->isConfigurated()))
-	  {
-		  static_cast<RuntimeType*>(this)->setConfiguration(configuration);
-	  }
-
-  }
-  
-  void startRT()
-  {
-        static_cast<RuntimeType*>(this)->start();
-  }
-
-  void removeObject(IStateMachine* sm)
-  {
-      static_cast<RuntimeType*>(this)->removeObject(sm);
-  }
-
-
-  void stopUponCompletion()
-  {
-        static_cast<RuntimeType*>(this)->stopUponCompletion();
-  }
+	/*!
+	Returns the runtime instance during the model execution.
+	*/
+	static ES::RuntimePtr<RuntimeType> getRuntimeInstance()
+	{
+		if (instance == nullptr)
+		{
+			instance = RuntimeType::createRuntime();
+		}
+		return instance;
+	}
+
+	/*!
+	Register a state machine for the runtime instance so
+	the threaded runtime can record the number of objects instances during the model execution.
+	Called by the state machine constructor.
+	*/
+	void setupObject(ES::StateMachineRef sm)
+	{
+		static_cast<RuntimeType*>(this)->setupObjectSpecificRuntime(sm);
+	}
+
+
+	/*!
+	Remove a state machine for the runtime instance when
```

Comment:
Should be _Removes_. Also sounds strange, I recommend _from_ instead of _for_. _When_ also doesn't feel appropriate here, but I don't have related suggestions, as I don't fully understand what's happening here.